doc/developer/arrangements.md

```diff
@@ -0,0 +1,98 @@
+
+To investigate the existing arrangements, query the `mz_arrangement_sizes` logging source.
+There are several diagnostic views in `mz_catalog` that connect this information to operator names, and group it by dataflow.
+
+## Which operators need Arrangements?
+
+We create more arrangements that just those to house materialized sources and views.
+Many internal operators require arrangements so that they can respond efficiently to changes.
+For example, the differential dataflow `join` operator requires that each of its two inputs be an arrangement.
+Additionally, the differential dataflow `reduce` operator requires that both its input and its output be an arrangement.
+These two operators are at the heart of the arrangement use in Materialize, because we construct relational operators out of them.
+
+Materialize has several relational operators that are "rendered" using differential `join` and `reduce` operators.
+These are the operators you would see in `explain plan for <query>`.
+There is a more complicated mapping from SQL to these operators.
+
+---
+
+Our `Join` operator joins together a sequence of input relations.
+It creates a sequence of differential `join` operators, each of which need to have both inputs be arrangements.
+This means that in addition to arranging each of the join inputs, it arranges each of the intermediate results.
+For example, in the figure below, each `A` is an arrangement.
+
+```
+    In2 --> A  In3 --> A  In4 --> A
+             \          \          \
+    In1 --> A-Join --> A Join --> A-Join --> Out
+```
+
+To join four inputs together, there are six arrangements.
+Some of these input arrangements could be shared, but most likely the intermediate arrangements cannot.
+
+---
+
+Our `Reduce` operator builds a non-trivial dataflow subgraph (which may be subject to change).
+Informally, for aggregation we build its own dataflow.
+That dataflow may begin with a `reduce` operator to implement the `distinct` keyword.
+Then there is either
+    * a single `reduce` operator (for sums, counts, any, all), or
+    * a sequence of 16 `reduce` operators (for min and max), or
+    * a single `reduce` operator (for `jsonbagg` and others).
+Finally, last `reduce` collects each of the aggregates into one record for each key.
+There are special cases when there are few (zero or one) reductions.
+
+---
+
+Our `TopK` operator builds a sequence of 16 `reduce` operators.
+Each `reduce` operator applies the `TopK` logic over progressively coarser groups
+The final group is just by the group key.
+
+---
+
+Our `Threshold` operator creates one `reduce` operator.
+
+---
+
+Our `ArrangeBy` operator creates one arrangement.
+
+---
+
+For each of these relational operators, the associated memory footprint should determined by the sizes of the arranged collections.
+
+## Caveats: Shared Arrangements
+
+Arrangements may be shared between operators in the same dataflow, or across dataflows for published arrangements.
```

Comment:
what is a "published arrangement"? one reading is that this is something that's explicitly a named index but maybe thats wrong and its any constructed arrangement?

Reply:
It's meant to mean "arrangements that are explicitly exported from a dataflow". I think at the moment this is just materialized sources, materialized views, and created indexes.